### PR TITLE
tiltfile: object identifiers are not case sensitive

### LINKED ISSUE #3423: k8s_resource object selector error message makes no sense
I'm trying to use the k8s_resource(objects=) syntax and getting this error message:

```
No object identified by the fragment "queue-proxy:Image:knative-serving" could be found. Possible objects are: ... "queue-proxy:Image:knative-serving", ...
```

but...that exact object is in the list? What's going on?

Here's the snapshot:
https://cloud.tilt.dev/snapshot/AcKL8-0LhBMFIMSeWiw=

cc @jazzdan do you have any ideas on this?

### PR DESCRIPTION
**Pull request by jazzdan**
Fixes #3423

### CODE REVIEW

File: internal/tiltfile/tiltfile_state.go [hunk 1]
```diff
@@ -655,7 +655,7 @@ func (s *tiltfileState) assembleK8sV2() error {
 			}
 
 			for i, o := range opts.objects {
-				entities, ok := fragmentsToEntities[o]
```
Comment (nicks) [inline internal/tiltfile/tiltfile_state.go:658]: i'm surprised they're not stored in objects as lowercase?

File: internal/tiltfile/tiltfile_state.go [hunk 2]
```diff
@@ -1331,28 +1331,44 @@ type k8sObjectSelector struct {
...
 // If an arg is an empty string it will become an empty regex that matches all input
 // Otherwise the arg must match the input exactly
-func newExactK8sObjectSelector(apiVersion string, kind string, name string, namespace string) (k8sObjectSelector, error) {
-	ret, err := newK8sObjectSelector(
-		exactOrEmptyRegex(apiVersion),
-		exactOrEmptyRegex(kind),
-		exactOrEmptyRegex(name),
-		exactOrEmptyRegex(namespace),
-	)
+func newExactCaseInsensitiveK8sObjectSelector(apiVersion string, kind string, name string, namespace string) (k8sObjectSelector, error) {
```
Comment (nicks) [inline internal/tiltfile/tiltfile_state.go:1334]: I feel like the nomenclature here needs to be cleaned up a bit, in particular:
- All functions are case insensitive now, but only one of them has "CaseInsensitive" in the name. That seems weird?
- The usage of the word "exact" here is not something i've seen before. I think python calls this "FullMatch"?

File: internal/tiltfile/tiltfile_state.go [hunk 3]
```diff
@@ -1331,28 +1331,44 @@ type k8sObjectSelector struct {
...
-	ret.namespaceString = namespace
+	ret.namespace, err = regexp.Compile(exactOrEmptyRegex(namespace))
+	if err != nil {
+		return k8sObjectSelector{}, errors.Wrap(err, "error parsing namespace regexp")
+	}
 
 	return ret, nil
 }
 
+func makeCaseInsensitive(s string) string {
```
Comment (nicks) [inline internal/tiltfile/tiltfile_state.go:1361]: i feel like half the problem here is that we're passing around a regexp around as a DSL string, rather than something more structured, and i wonder if this code would be easier to follow if we had something like

```
type re struct {
  pattern string
  ignoreCase bool
  fullMatch bool
}

func (r re) compile() { ... }
```

File: internal/tiltfile/tiltfile_state.go (hunk 2 above)
Comment (jazzdan) [reply internal/tiltfile/tiltfile_state.go:1334]: ooh I like FullMatch. And I agree, I'll clean this stuff up.

File: internal/tiltfile/tiltfile_state.go (hunk 3 above)
Comment (jazzdan) [reply internal/tiltfile/tiltfile_state.go:1361]: Yeahhh I tried something like this back when I was originally writing this a couple weeks ago. I'll give it another shot with fresh eyes.

File: internal/tiltfile/tiltfile_state.go (hunk 1 above)
Comment (jazzdan) [reply internal/tiltfile/tiltfile_state.go:658]: Hmm it might be weird to use a different string than the one the user entered in error messages though, even if we treat it as case insensitive. I think I want to preserve the original string.

**Comment by landism** [comment]
Did you end up finding a source or some evidence for whether k8s allows two objects to have the same name with different casing? It's probably worth writing down the evidence/assumption, given we were unclear on it.

**Comment by nicks** [comment]
Here are some references:
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#api-conventions
https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names

As a general rule of thumb, Kubernetes tends to convert everything to lowercase, since they end up getting converted to case-insensitive HTTP endpoints

**Comment by jazzdan** [comment]
Yeah that stuff is what I went off of.

I'm still a little unclear on kind names, but it seems that most everything has been lowercase in my testing.

Like in this part of the docs

>Resource collections should be all lowercase and plural, whereas kinds are CamelCase and singular. Group names must be lower case and be valid DNS subdomains.

It doesn't explicitly say that kind names are case insensitive, but I wasn't able to test this myself yet.

**Comment by nicks** [comment]
Ya, I think the API docs are being super-precise in a way that's helpful if you understand the internals, but hard to parse if you do not.

There are two different things:
- kinds (which require camel case)
- resources (which are exposed as lowercase)
and there's a mapping from Kind -> resource that converts to lowercase

but to a user (e.g., someone who's just using kubectl and writing yaml), because there's a 1<>1 mapping, it's easy to read that as "oh, it's case-insensitive", when what's actually happening is that you're interacting with different things.

Does that make sense?

**Comment by jazzdan** [comment]
Yeahhh that makes sense. Thanks @nicks!

**Comment by jazzdan** [comment]
I updated some nomenclature but left the refactor as a TODO because, once again, I ran in to weird test failures. I think we're depending on some regex behavior that is unclear to me and super finicky.
